compiler/src/parsing/ast_helper.re

```diff
@@ -384,25 +390,111 @@ module Expression = {
             tuple_construct(~loc, ~attributes?, ident_cons, [expr, empty])
           | ListSpread(expr, _) => expr
           };
-        List.fold_left(
-          (acc, expr) => {
+        if (List.exists(
+              expr =>
+                switch (expr) {
+                | ListSpread(_) => true
+                | ListItem(_) => false
+                },
+              rest,
+            )) {
+          collection_concat(
+            ~loc,
+            ~attributes?,
+            PExpListConcat,
+            List.map(
+              expr =>
+                switch (expr) {
+                | ListSpread(expr, loc) => (
+                    PExpSpreadExpr,
+                    {...expr, pexp_loc: loc},
+                  )
+                | ListItem(expr) => (
+                    PExpNonSpreadExpr,
+                    // Still convert to a single-element list to make later compilation steps easier
```

Comment:
This seems a little unnecessary. To me this looks like you're converting `[a, b, ...c, d, e, ...f]` into `concat([> [a], [b], c, [d], [e], f])`, instead of `concat([> [a, b, ...c], [d, e, ...f]])`.

Reply:
True, but I think the overall runtime complexity would be the same since appending a single-element list on the left and appending an element to an existing list are both constant-time

Reply:
Complexity-wise yes, but you get two allocations for a single-item list (cons cell + empty) versus one (just a cons cell).